checkbox-ng/plainbox/impl/ctrl.py

```diff
@@ -125,15 +128,64 @@ def get_dependency_set(self, job):
             resource_deps = job.get_resource_dependencies()
         except ResourceProgramError:
             resource_deps = ()
+        suspend_job_id_list = [
+            Suspend.AUTO_JOB_ID,
+            Suspend.MANUAL_JOB_ID,
+        ]
+        if job.id in suspend_job_id_list:
+            suspend_deps = self._get_before_suspend_dependency_set(
+                job.id, job_list
+            )
+        else:
+            suspend_deps = set()
         result = set(
             itertools.chain(
                 zip(itertools.repeat(direct), direct_deps),
                 zip(itertools.repeat(resource), resource_deps),
                 zip(itertools.repeat(ordering), after_deps),
+                zip(itertools.repeat(ordering), suspend_deps),
             )
         )
         return result
 
+    def _get_before_suspend_dependency_set(self, suspend_job_id, job_list):
+        """
+        Get the set of after dependencies of a suspend job.
+
+        Jobs that have a ``also-after-suspend[-manual]`` flag should be run
+        before their associated suspend job. Similary, jobs that declare a
+        sibling with a dependency on a suspend job should be run before said
+        suspend job. This function finds these jobs and add them as a
+        dependency for their associated suspend job.
+
+        :param suspend_job_id:
+            The id of a suspend job. One of the following is expected:
+            Suspend.AUTO_JOB_ID or Suspend.MANUAL_JOB_ID.
+        :param job_list:
+            List of jobs to search dependencies on.
+        :returns:
+            A set of job ids that need to be run before the suspend job
+        """
+        suspend_deps = set()
+        expected_flag = ""
+        if suspend_job_id == Suspend.AUTO_JOB_ID:
+            expected_flag = Suspend.AUTO_FLAG
+        elif suspend_job_id == Suspend.MANUAL_JOB_ID:
+            expected_flag = Suspend.MANUAL_FLAG
+        if not expected_flag:
+            return suspend_deps
+        for dep_job in job_list:
```

Comment:
If you create a function that takes one job (as in one item in the dep_job collection), and do same check in that function, you can then run return set(filter(the_predicate_fn, job_list)).
Testability and test clarity being the big benefit with that.

Reply:
This twisted my brain for a few hours today, but I think I managed to get something to work along these lines. Thanks for the feedback, TIL!